swayfmt/src/utils/map/comments.rs

```diff
@@ -232,15 +232,22 @@ fn insert_after_span(
     let iter = comments_to_insert.iter();
     let mut offset = offset;
     let mut comment_str = String::new();
+    let mut pre_module_comment = false;
     for comment_with_context in iter {
         let (comment_value, comment_context) = comment_with_context;
+        if comment_value.span.start() == 0 {
```

Comment:
I wonder if this condition should be:

```suggestion
        if comment_value.span.start() < from.start() {
```

Though I'm not sure exactly what the `from` span represents here - is it the span of the current item being checked for comments?

---

The issue that comes to mind with using `0` is that this might not work in the case that there's some empty whitespace at the top of the file, e.g.

```rust
 
// Blah blah
library foo;
```
Arguably, there shouldn't be any unnecessary leading whitespace at the top of the file, though we can't be sure the user won't run the formatter with an extra newline at the top.

One potentially hacky solution might be to always `.trim()` the unformatted code `str` before we begin any formatting or parsing to make sure there's no empty whitespace at the beginning or end of the `str`. I think in that case, the comparison to `== 0` here should be OK as if there is a comment preceding the module kind, it would always start at `0`. We might want to add a test to check the case of having an extra newline before the first comment just in case.

Reply:
This is actually something I was thinking as well, though in a slightly different context. The comments I think should never have a preceding new line, all new lines should be handled in `newline_handling` including ones that are on comments.

Reply:
#2790